Describe uncaught error incident handling (#282)
camunda-cloud/camunda-cloud-documentation: docs/reference/bpmn-processes/error-events/error-events.md

title: "Error events"
---

Error events are events which reference an error.

> They are used to handle business errors in a process.

Author: I'm not sure this needs an annotation. In my opinion, it's more clearly described in the throwing the error section now. WDYT?

![process](assets/error-events.png)

An error indicates that some kind of business error has occurred which should be handled in the process, for example, by taking a different path to compensate the error.

## Defining the error

> An error can be referenced by one or more error events.

Author: I think this originates from the bpmn spec, where the error is defined in the xml and can be referenced from multiple error events. But we can improve the wording to make it more clear what is meant.

> It must define the `errorCode` (e.g. `Invalid Credit Card`) of the error.

Reviewer: @pihme yes but it should still be clear that an error code is mandatory.

The `errorCode` is a `string` that must match to the error code that is sent by the client command or from the error end event.

## Throwing the error

> You can throw an error from a **client command** while processing a job.

Reviewer 2: this section might be easier to read if triggers and effects are split:

The error can be thrown while processing a job. Or it can be thrown inside a process using the end event.

Image

Errors thrown in job disable the job and ...
Errors thrown in end events do not stop the execution of other branches in the same process. However, the error event "bubbles up" through the scopes until it is caught.

Which also gives a transition to the next section

Author: This was difficult, so I took another route. WDYT about it now?

title: "Error events"
---

> In process automation, you may have to handle with errors related to the process, also known as business errors.

Reviewer 2: ```suggestion
In process automation, you may have to deal with errors related to the process, also known as business errors.
```

Author: I had this at some point because I wanted to introduce the term `error handling` earlier, but I struggled with the difference between handle and catch. I've talked with Philipp and we think it's a good idea to move away from the word handle altogether. I'll only leave it in the heading `Unhandled error` because the incident (due to not catching the error) has the unhandled error type. Everything else relates to catching which is also the term used by the bpmn spec.

Some examples of business errors are: an invalid credit card used as default payment method to collect money, or cancelling an order after it was already payed for.
In case of a business error, you may need to take a different path than usual or compensate for the error.

![process](assets/error-events.png)

## Defining the error

In BPMN, business errors are represented as errors and error events.
Errors define the possible errors that can occur.
Error events are elements in the process that refer to the defined errors.
An error can be referenced by one or more error events.

An error must define an `errorCode` (e.g. `Invalid Credit Card`).
The `errorCode` is a `string` that is used to match a thrown error to the error catch events.

## Throwing the error
An error can be thrown from within the process using an error **end event**.

![process](assets/error-throw-events.png)

Alternatively, you can inform Zeebe that a business error has occurred using a **client command**.

> As a client command an error can be thrown only while processing a job.

Reviewer 2: ```suggestion
This client command can only be thrown while processing a job.
```

Author: ```suggestion
This client command can only be used to throw an error while processing a job.
```

In addition to throwing the error, this also disables the job and stops it from being activated or completed by other job workers.
See the [gRPC command](/reference/grpc.md#throwerror-rpc) for details.

## Catching the error

A thrown error can be caught by an error catch event.
Specifically, using an error **boundary event** or an error **event subprocess**.

Starting at the scope where the error was thrown, the error code is matched against the attached error boundary events and error event sub processes at that level.
An error is caught by the first event in the scope hierarchy that matches the error code.

> At each scope, the error is propagated to the parent scope if it is not yet caught.

Reviewer 2: ```suggestion
At each scope, the error is either caught, or it is propagated to the parent scope.
```

In case the process instance is created via call activity, the error can also be caught in the calling parent process instance.

Error boundary events and error event subprocesses must be interrupting.
This means the process instance will not continue along the regular path when the error was caught.
Instead, it continues where the error is caught.
If the error is thrown for a job, then the associated task gets terminated first.
To continue the execution, the error boundary event or error event subprocess that caught the error gets activated.

## Unhandled errors

When an error is thrown then it should be caught.
If it is not caught, then an **incident** (i.e. `Unhandled error event`) is raised to indicate the failure.
The incident is attached to the corresponding element where the error was thrown, i.e. the task of the processed job or the error end event.

> When you resolve the incident attached to a task, it will ignore the error, re-enable the job and allow it to be activated and completed by a job worker once again.

Reviewer: (optional) We could improve the wording a bit. The part "it will ignore the error" could be misleading.

  ```suggestion
Resolving the incident attached to a task will re-enable the job and allow it to be activated and completed by a job worker once again.
```

Author: How do you feel about `it will swallow the error`? Point is that the error will not be caught anymore. It was thrown, but we just ignore it.

Reviewer: I don't know :sweat_smile: I find this part more confusing than helpful. But it could be just me. Feel free to merge it as it is :+1:
